Работа с проектами. Проекты.

Edisolt5498 opened the pull request:
Проекты

Commit:
init commit. Now we can create project with validation. No tests and handlers yet.

Commit:
filters with generic added. Also new endpoints and handler

Commit:
All done. Just need to add tests and handlers

Commit:
All done. Just need to add tests and handlers

Commit:
Tests added. getAll() and getAllWithFilters() tests need to be finished.

Commit:
Done!

Commit:
Done! Validator is component now!

Commit:
Done! Validator is component now!

Commit:
validation .validateUserExistence is in ProjectValidator also now.

Commit:
validation .validateUserExistence is in ProjectValidator also now.

InnMish commented on src/main/java/faang/school/projectservice/contreoller/project/ProjectController.java:
```diff
@@ -0,0 +1,56 @@
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/project-service/api/v1/projects")
```
сначала api v1, потом уже детали 
project-service не нужно, потому что это наше внутреннее название репозитория, пользователю незачем знать в какой репозиторий наш он отправляет запрос) 

Далее: знаю, что это апи в1 из лекции, но есть, разумеется, хорошая практика 

Общие вот такие моменты для всех контроллеров (а апи в1 - это версия нашей апишки), выносим в конфиг-файл 
В нашей случае это application yaml 
так, здесь у тебя останется только projects, а все общее будет вынесено в конфиг

InnMish commented on src/main/java/faang/school/projectservice/contreoller/project/ProjectController.java:
```diff
@@ -0,0 +1,56 @@
+    @GetMapping("/test")
+    public String test() {
+        return String.valueOf(userContext.getUserId());
+    }
```
это еще кто такой)

InnMish commented on src/main/java/faang/school/projectservice/contreoller/project/ProjectController.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    @GetMapping("/filters")
+    public List<ProjectDto> getByFilters(@ModelAttribute ProjectFilterDto filterDto) {
```
@ModelAttribute достает параметры из html формы, насколько я знаю 
А как у нас тут работает? 
Как RequestBody?

Edisolt5498 replied:
Да, как RequestBody.
Я хотел, чтоб endpoint фильтра работал через параметры, а не тело запроса. А в таком случае приходилось вручную заполнять Dto
Нашел такую штуку (ModelAttribute ), заполняет Dto в зависимости от переданных параментров в данном случае

Edisolt5498 replied:
А, стоп, так именно это ты и сказала. -_- Да, достает параметры из формы)

InnMish commented on src/main/java/faang/school/projectservice/handler/ControllerError.java:
```diff
@@ -0,0 +1,19 @@
+@Setter
+@ToString
+@AllArgsConstructor
+public class ControllerError {
```
Контроллер еррор говорит как будто бы о том, что ошибки тут только из контроллера)
Лучше назвать ResponseError, например

InnMish commented on src/main/java/faang/school/projectservice/service/project/ProjectService.java:
```diff
@@ -0,0 +1,99 @@
+        String name = projectDto.getName();
+        String description = projectDto.getDescription();
+
+        projectValidator.validateAccessToProject(ownerId);
+        projectValidator.validateUserExistence(ownerId);
+        projectValidator.validateNameExistence(ownerId, name);
+        projectValidator.validateName(name);
+        projectValidator.validateDescription(description);
```
вот эту прелесть внутри ProjectValidator объедини в один общий метод validate  
Пусть там внутри они делятся на мелкие методы, а тут будет один вызов всего сразу

InnMish commented on src/main/java/faang/school/projectservice/service/project/ProjectService.java:
```diff
@@ -0,0 +1,99 @@
+    public ProjectDto create(ProjectDto projectDto) {
+        long ownerId = projectDto.getOwnerId();
+        String name = projectDto.getName();
+        String description = projectDto.getDescription();
```
если поле только один раз используется, можно не выносить его в поле отдельное

InnMish commented on src/main/java/faang/school/projectservice/service/project/ProjectService.java:
```diff
@@ -0,0 +1,99 @@
+        projectValidator.validateDescription(description);
+
+        projectDto.setStatus(ProjectStatus.CREATED);
+        if (projectDto.getVisibility() == null) {
```
а проверь в Entity Project, есть ли у этого поля значение по умолчание? Там аннотация должна висеть на поле такая
Если не висит, то залезь в миграцию (resources -> db -> и там найти нужный файлик, где создается project таблица) и там посмотреть, есть ли значение по умолчанию у этого поля. Если нет, то ок, все нормально тут

Edisolt5498 replied:
На сколько я понимаю, нет дефолтных значений
entity:
@Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private ProjectStatus status;

db:
status VARCHAR(255) NOT NULL,

InnMish commented on src/main/java/faang/school/projectservice/service/project/ProjectService.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    public ProjectDto update(ProjectDto projectDto) {
+        Project project = projectRepository.getProjectById(projectDto.getId());//also validation - throws if no project
```
отдельный метод надо сделать для поиска по айди, он у тебя много раз используется (как минимум два :) ) 
вообще я что-то не поняла, почему тут у нас прям проект возвращается, а не optional  - но это в репозитории видимо так

Edisolt5498 replied:
Репа выполняет эту функцию, она там получает Optional и бросает ошибку при null. Не знаю почему так, могу поменять

InnMish commented on src/main/java/faang/school/projectservice/service/project/ProjectService.java:
```diff
@@ -0,0 +1,99 @@
+        ProjectStatus status = projectDto.getStatus();
+        String description = projectDto.getDescription();
+
+        if (status != null) {
+            project.setStatus(status);
+        }
+        if (description != null) {
```
а если null, не будем сетить? 
Я бы засетила

Edisolt5498 replied:
Типа, если нам не передали параметр, то не обновляем его, нет?

InnMish commented on src/main/java/faang/school/projectservice/service/project/ProjectService.java:
```diff
@@ -0,0 +1,99 @@
+            projectValidator.validateDescription(description);
+            project.setDescription(description);
+        }
+        project.setUpdatedAt(LocalDateTime.now());
```
это на уровне базы автоматически должно проставляться 
Если в энтити над полем висит аннотация типа UpdatedTimestamp или какая-то похожая, значит, руками не надо сетить поле

InnMish commented on src/main/java/faang/school/projectservice/validator/project/ProjectValidator.java:
```diff
@@ -0,0 +1,52 @@
+
+    public void validateName(String name) {
+        if (name.isEmpty() || name.isBlank()) {
+            throw new IllegalArgumentException("Name of project cannot be empty or blank");
```
dataValidationException

InnMish commented on src/main/java/faang/school/projectservice/validator/project/ProjectValidator.java:
```diff
@@ -0,0 +1,52 @@
+
+    public void validateDescription(String description) {
+        if (description != null && (description.isEmpty() || description.isBlank())) {
+            throw new IllegalArgumentException("Description of project cannot be empty or blank");
```
dataValidationException

InnMish commented on src/main/java/faang/school/projectservice/validator/project/ProjectValidator.java:
```diff
@@ -0,0 +1,52 @@
+
+    public void validateNameExistence(long ownerId, String name) {
+        if (projectRepository.existsByOwnerUserIdAndName(ownerId, name)) {
+            throw new IllegalArgumentException("Project with this name already exists. Name: " + name);
```
EntityNotFoundException

Edisolt5498 replied:
Так там же ошибка о том, что проект уже есть

InnMish commented on src/main/java/faang/school/projectservice/validator/project/ProjectValidator.java:
```diff
@@ -0,0 +1,52 @@
+
+    public void validateUserExistence(long ownerId) {
+        UserDto user = userServiceClient.getUser(ownerId);
+        if (user == null) {
+            throw new EntityNotFoundException("User with id = " + ownerId + " not found");
```
А такое может произойти? Тестировал? 
По идее если ты обратишься в user_service за пользователем здесь, то именно метод в us тебе кинет ошибку о том, то пользователь не найден 
И сюда null уже не придет

Edisolt5498 replied:
Просто у нас нет пока endpointa под это дело, поэтому так оставил пока. Но это конечно так должно быть, согласен

Commit:
/api/v1 in config
fixed bag with access to not own project
naming of exceptions
new handler for MethodArgumentNotValidException
refactor

InnMish commented on src/main/java/faang/school/projectservice/contreoller/project/ProjectController.java:
```diff
@@ -0,0 +1,52 @@
+@RestController
+@Validated
+@RequiredArgsConstructor
+@RequestMapping("${endpoint_base.path}/projects")
```
Нет, не так) 
Есть прям спринговая настройка, в которой можно указать пусть 
там что-то типа spring.mvc...  нужно погуглить, как в ямле прописать путь 

Тогда тебе сюда вообще не надо будет этот параметр прокидывать, он по умолчанию применится

InnMish commented on src/main/java/faang/school/projectservice/service/project/ProjectService.java:
```diff
@@ -0,0 +1,96 @@
+
+    public ProjectDto create(ProjectDto projectDto) {
+        projectValidator.validateToCreate(projectDto);
+        //userServiceClient.getUser(projectDto.getOwnerId()); //throws if user doesn't exist
```
почему коммент? Это нужно починить и сделать так, чтобы работало в связке с us

Commit:
/api/v1 in config
fixed userClient

Commit:
refactor userClient
added handler for feignClients

Commit:
refactor userClient
added handler for feignClients
